File: src/engine/reporter.ts
// src/engine/reporter.ts
import type { AuditResult, CheckResult, Finding } from "./types";

const STATUS_LABEL: Record<string, string> = {
  pass: "PASS  ",
  fail: "FAIL  ",
  exempt: "EXEMPT",
  skip: "SKIP  ",
  warn: "WARN  ",
};

export interface ReporterOptions {
  verbose?: boolean;
}

function formatFinding(f: Finding): string {
  const loc = f.file
    ? f.line
      ? `${f.file}:${f.line}`
      : f.file
    : "";
  const sev = f.severity.padEnd(7);
  return `  ${sev}  ${loc.padEnd(30)}  ${f.message} (${f.ruleId})`;
}

export function printAuditResult(result: AuditResult, options?: ReporterOptions): void {
  console.log("\n=== agent-config audit ===\n");

  const byContract = new Map<string, CheckResult[]>();
  for (const r of result.results) {
    const existing = byContract.get(r.contractId) ?? [];
    existing.push(r);
    byContract.set(r.contractId, existing);
  }

  for (const [id, checks] of byContract) {
    const worst = checks.find(c => c.status === "fail")
      ?? checks.find(c => c.status === "warn")
      ?? checks[0]!;

    const label = STATUS_LABEL[worst.status] ?? worst.status.toUpperCase().padEnd(6);
    const desc = checks[0]!.contractDescription;

    // Collect all findings across checks for this contract
    const allFindings = checks.flatMap(c => c.findings ?? []);
    const findingCount = allFindings.length;

    const extra = findingCount > 0
      ? `  — ${findingCount} finding${findingCount !== 1 ? "s" : ""}`
      : worst.message
        ? `  — ${worst.message}`
        : "";

    console.log(`${id.padEnd(12)} ${label}  ${desc}${extra}`);

    if (options?.verbose && findingCount > 0) {
      for (const f of allFindings) {
        console.log(formatFinding(f));
      }
    }
  }

  const { passed, failed, exempt, skipped, warned } = result;
  console.log(
    `\n=== ${passed} passed, ${failed} failed, ${exempt} exempt, ${skipped} skipped, ${warned} warned ===\n`
  );
}

export function formatAuditResultJson(result: AuditResult): string {
  return JSON.stringify(result, null, 2);
}

File: internal/engine/reporter.go
package engine

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

var statusLabel = map[CheckStatus]string{
	StatusPass:   "PASS  ",
	StatusFail:   "FAIL  ",
	StatusWarn:   "WARN  ",
	StatusExempt: "EXEMPT",
	StatusSkip:   "SKIP  ",
	StatusHalt:   "HALT  ",
}

// FormatText returns a human-readable audit report.
// verbose=true shows per-file results; false collapses to per-contract.
func FormatText(result AuditResult, verbose bool) string {
	var b strings.Builder
	b.WriteString("\n=== contracts check ===\n\n")

	// Group by contract: worst status wins
	seen := map[string]CheckStatus{}
	msgs := map[string]string{}
	for _, r := range result.Results {
		prev, ok := seen[r.ContractID]
		if !ok {
			seen[r.ContractID] = r.Status
			msgs[r.ContractID] = r.ContractDescription
			if r.Message != "" {
				msgs[r.ContractID] += "  — " + r.Message
			}
			continue
		}
		if worse(r.Status, prev) {
			seen[r.ContractID] = r.Status
			if r.Message != "" {
				msgs[r.ContractID] = r.ContractDescription + "  — " + r.Message
			}
		}
	}

	ids := make([]string, 0, len(seen))
	for id := range seen {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	for _, id := range ids {
		label := statusLabel[seen[id]]
		fmt.Fprintf(&b, "%s  %s  %s\n", id, label, msgs[id])
	}

	summary := fmt.Sprintf("%d passed, %d failed, %d warned, %d exempt, %d skipped",
		result.Passed, result.Failed, result.Warned, result.Exempt, result.Skipped)
	if result.Halted > 0 {
		summary += fmt.Sprintf(", %d halted", result.Halted)
	}
	fmt.Fprintf(&b, "\n=== %s ===\n", summary)
	return b.String()
}

func worse(a, b CheckStatus) bool {
	rank := map[CheckStatus]int{
		StatusHalt: 5, StatusFail: 4, StatusWarn: 3, StatusExempt: 2, StatusSkip: 1, StatusPass: 0,
	}
	return rank[a] > rank[b]
}

// FormatJSON returns the audit result as pretty-printed JSON.
func FormatJSON(result AuditResult) string {
	out, _ := json.MarshalIndent(result, "", "  ")
	return string(out)
}
